### PR TITLE
Fix NFT rule/role update order

### PR DESCRIPTION
**Pull request by vimukthi-git**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Issue Link: #519

### CODE REVIEW

File: nft/ethereum_payment_obligation.go [hunk 1]
```diff
@@ -143,79 +129,113 @@ func (s *ethereumPaymentObligation) MintNFT(ctx context.Context, documentID []by
...
-		log.Infof("Sent off ethTX to mint [tokenID: %x, anchor: %x, registry: %x] to payment obligation contract. Ethereum transaction hash [%x] and Nonce [%v] and Check [%v]",
-			requestData.TokenID, requestData.AnchorID, requestData.To, ethTX.Hash(), ethTX.Nonce(), ethTX.CheckNonce())
-		log.Infof("Transfer pending: 0x%x\n", ethTX.Hash())
+		model, err = s.docSrv.DeriveFromCoreDocument(cd)
+		if err != nil {
+			errOut <- err
+		}
 
-		res, err := ethereum.QueueEthTXStatusTask(accountID, txID, ethTX.Hash(), true, s.queue)
+		_, _, err = s.docSrv.Update(contextutil.WithTX(ctx, txID), model)
```
Comment (vimukthi-git) [inline nft/ethereum_payment_obligation.go]: Doing this rule update and anchoring before minting is the main change

File: nft/payment_obligation.go [hunk 2]
```diff
@@ -2,8 +2,42 @@ package nft
...
+
+// BigInt converts tokenID to big int
+func (t TokenID) BigInt() *big.Int {
+	return utils.ByteSliceToBigInt(t[:])
+}
+
+// URI gets the URI for this token
+func (t TokenID) URI() string {
+	// TODO please fix this
+	return "http:=//www.centrifuge.io/DUMMY_URI_SERVICE"
```
Comment (lucasvo) [inline nft/payment_obligation.go]: Will you fix this? I think here we should read from the ethereum contract to get the URI. This should be part of the NFT epic.

File: nft/ethereum_payment_obligation.go (hunk 1 above)
Comment (manuelpolzhofer) [reply nft/ethereum_payment_obligation.go]: As discussed in person. We need to wait for the document update because the latest current version is needed and the tokenID needs to be part of the merkle proof.

File: nft/ethereum_payment_obligation.go [hunk 3]
```diff
@@ -111,111 +115,149 @@ func (s *ethereumPaymentObligation) prepareMintRequest(ctx context.Context, docu
...
 
-// sendMintTransaction sends the actual transaction to mint the NFT
-func (s *ethereumPaymentObligation) sendMintTransaction(
-	cid identity.CentID,
-	contract ethereumPaymentObligationContract,
-	opts *bind.TransactOpts,
-	requestData MintRequest,
-	registry common.Address,
-	documentID []byte) (uuid.UUID, error) {
+		cd, err := model.PackCoreDocument()
```
Comment (manuelpolzhofer) [inline nft/ethereum_payment_obligation.go]: After the model refactoring, we can call the model directly here correct?

File: transactions/manager.go [hunk 4]
```diff
@@ -12,38 +12,69 @@ import (
...
+type Manager interface {
 	// ExecuteWithinTX executes a transaction within a transaction
 	ExecuteWithinTX(ctx context.Context, accountID identity.CentID, existingTxID uuid.UUID, desc string, work func(accountID identity.CentID, txID uuid.UUID, txMan Manager, err chan<- error)) (txID uuid.UUID, done chan bool, err error)
-
-	// CreateTransaction
-	// TODO [TXManager] remove this once TX Manager update is complete
-	CreateTransaction(accountID identity.CentID, desc string) (*Transaction, error)
 	GetTransaction(accountID identity.CentID, id uuid.UUID) (*Transaction, error)
-	SaveTransaction(tx *Transaction) error
+	UpdateTaskStatus(accountID identity.CentID, id uuid.UUID, status Status, taskName, message string) error
```
Comment (vimukthi-git) [inline transactions/manager.go]: This is a key txManager update in this pr, we use task status instead of tx status to track progress now in tasks

File: documents/anchor_task.go [hunk 5]
```diff
@@ -119,31 +119,41 @@ func (d *documentAnchorTask) RunTask() (res interface{}, err error) {
...
+	tr, err := tq.EnqueueJob(documentAnchorTaskName, params)
+	if err != nil {
+		return nil, err
+	}
+
+	return tr, nil
+}
+
+// CreateAnchorTransaction creates a transaction for anchoring a document using transaction manager
+func CreateAnchorTransaction(txMan transactions.Manager, tq queue.TaskQueuer, self identity.CentID, txID uuid.UUID, documentID []byte) (uuid.UUID, chan bool, error) {
```
Comment (vimukthi-git) [inline documents/anchor_task.go]: this would be used by all doc updates to create or update an existing tx with task status.

Comment (vimukthi-git) [reply documents/anchor_task.go]: need to add task status as pending here

**Comment by vimukthi-git** [comment]
@manuelpolzhofer yes those are different parts of NFT uniqueness, that we need to implement. This one is only about getting the flow right to provide proofs of uniqueness to the contract.

File: nft/ethereum_payment_obligation.go [hunk 6]
```diff
@@ -111,111 +115,149 @@ func (s *ethereumPaymentObligation) prepareMintRequest(ctx context.Context, docu
...
+	}, done, nil
 }
 
-// OwnerOf returns the owner of the NFT token on ethereum chain
-func (s *ethereumPaymentObligation) OwnerOf(registry common.Address, tokenID []byte) (owner common.Address, err error) {
-	contract, err := s.bindContract(registry, s.ethClient)
-	if err != nil {
-		return owner, errors.New("failed to bind the registry contract: %v", err)
-	}
+func (s *ethereumPaymentObligation) minter(ctx context.Context, tokenID TokenID, documentID []byte, registryAddress, depositAddress string, proofFields []string) func(accountID identity.CentID, txID uuid.UUID, txMan transactions.Manager, errOut chan<- error) {
```
Comment (manuelpolzhofer) [inline nft/ethereum_payment_obligation.go]: I would split the minter funcs maybe into smaller funcs.

File: documents/service.go [hunk 7]
```diff
@@ -58,10 +58,10 @@ type Service interface {
 	ReceiveAnchoredDocument(ctx context.Context, model Model, senderID []byte) error
 
 	// Create validates and persists Model and returns a Updated model
-	Create(ctx context.Context, model Model) (Model, uuid.UUID, error)
+	Create(ctx context.Context, model Model) (Model, uuid.UUID, chan bool, error)
```
Comment (vedhavyas) [inline documents/service.go]: not sure why we need to have the chan here since this is ignored mostly. I would prefer not having this part of the interface

File: nft/ethereum_payment_obligation.go (hunk 6 above)
Comment (vedhavyas) [reply nft/ethereum_payment_obligation.go]: cyclomatic complexity of this function is more than optimum. Why did we move this here instead of having it part of the documents package? I prefer NFT package doesn't know about the documents package. Mpve this to documents package pls

File: nft/ethereum_payment_obligation_test.go [hunk 8]
```diff
@@ -226,13 +195,13 @@ func TestPaymentObligationService(t *testing.T) {
...
 			ctxh := testingconfig.CreateAccountContext(t, &mockCfg)
-			_, err := service.MintNFT(ctxh, decodeHex(test.request.Identifier), test.request.RegistryAddress, test.request.DepositAddress, test.request.ProofFields)
+			_, _, err := service.MintNFT(ctxh, decodeHex(test.request.Identifier), test.request.RegistryAddress, test.request.DepositAddress, test.request.ProofFields)
 			if test.err != nil {
 				assert.Equal(t, test.err.Error(), err.Error())
 			} else if err != nil {
 				panic(err)
 			}
-			docService.AssertExpectations(t)
+			//docService.AssertExpectations(t)
```
Comment (vedhavyas) [inline nft/ethereum_payment_obligation_test.go]: uncomment or remove pls

File: nft/ethereum_payment_obligation_test.go [hunk 9]
```diff
@@ -226,13 +195,13 @@ func TestPaymentObligationService(t *testing.T) {
...
 			paymentOb.AssertExpectations(t)
 			idService.AssertExpectations(t)
 			//ethClient.AssertExpectations(t)
```
Comment (vedhavyas) [inline nft/ethereum_payment_obligation_test.go]: this one too, pls

File: documents/service.go (hunk 7 above)
Comment (vimukthi-git) [reply documents/service.go]: its needed so that any code that uses this service function can wait on the update in an async way or the proper go way, instead of polling db. We allow polling of API for outside users but no need to do that inside the node. Therefore this has to be a convention we follow for all aync background funcs.

File: nft/ethereum_payment_obligation.go (hunk 6 above)
Comment (vimukthi-git) [reply nft/ethereum_payment_obligation.go]: I thing this is refactoring we can consider when we have the entire flow ready, `premature optimisation is the root of all evil` :)

**Comment by vimukthi-git** [comment]
Merging this now, Follow ups - 
- Call smart contract to get the token URL
- Refactor `mint` closure method to be smaller once we have entire flow
- More tests for txManager edge cases

cc @lucasvo @manuelpolzhofer @vedhavyas

File: nft/ethereum_payment_obligation.go (hunk 6 above)
Comment (vedhavyas) [reply nft/ethereum_payment_obligation.go]: I don't consider this to be a premature optimization @vimukthi-git 
I'm talking about breaking down the function to smaller chunks so that its easier to test and understand

Comment (vimukthi-git) [reply nft/ethereum_payment_obligation.go]: Agree, just that it seems kind of arbitrary to break it down right now, as only testing we can do is to mock things out individually and see if the function uses mocks correctly. I would love to think about this separately and do a proper test when completing the flow.
